File: various/src/main/java/com/chrnie/various/DefaultItemMatcherFactory.java
package com.chrnie.various;

import android.support.v4.util.ArrayMap;
import android.util.SparseArray;
import java.util.List;

public final class DefaultItemMatcherFactory implements ItemMatcher.Factory {

  private static final DefaultItemMatcherFactory INSTANCE = new DefaultItemMatcherFactory();

  public static DefaultItemMatcherFactory getInstance() {
    return INSTANCE;
  }

  private DefaultItemMatcherFactory() {

  }

  @Override
  public ItemMatcher create(List<Item> itemList) {
    return new DefaultItemMatcher(itemList);
  }

  static final class DefaultItemMatcher implements ItemMatcher {

    private static final ArrayMap<Class, Integer> CLASS_OF_INDEX = new ArrayMap<>(8);

    private final SparseArray<Item> indexOfItem;

    DefaultItemMatcher(List<Item> itemList) {
      indexOfItem = new SparseArray<>(itemList.size());
      for (Item item : itemList) {
        Class clz = item.dateType;

        Integer index = CLASS_OF_INDEX.get(clz);
        if (index == null) {
          index = CLASS_OF_INDEX.size() + 1;
          CLASS_OF_INDEX.put(clz, index);
        }

        indexOfItem.put(index, item);
      }
    }

    @Override
    public int getViewType(Object date) {
      Class clz = date.getClass();

      Integer index = CLASS_OF_INDEX.get(clz);
      if (index == null) {
        throw new RuntimeException(
            String.format("%s not found match item, make sure it has been registered", clz.getName()));
      }

      return index;
    }

    @Override
    public Item getItem(int viewType) {
      return indexOfItem.get(viewType);
    }
  }
}

File: various/src/main/java/com/chrnie/various/DefaultItemMatcherFactory.kt
package com.chrnie.various

import android.util.SparseArray
import androidx.collection.ArrayMap
import androidx.recyclerview.widget.RecyclerView

object DefaultItemMatcherFactory : ItemMatcher.Factory {

    private val dataTypeToIndex = ArrayMap<Class<*>, Int>()

    override fun create(itemList: List<Item<*, *>>): ItemMatcher {
        val indexToBinder = SparseArray<ViewBinder<Any, RecyclerView.ViewHolder>>(itemList.size)
        itemList.forEach { (dataType, viewBinder) ->

            val index = (dataTypeToIndex[dataType]
                    ?: (dataTypeToIndex.size + 1).also { dataTypeToIndex[dataType] = it })

            indexToBinder.put(index, viewBinder as ViewBinder<Any, RecyclerView.ViewHolder>)
        }

        return DefaultItemMatcher(indexToBinder)
    }

    private class DefaultItemMatcher(
        private val indexToBinder: SparseArray<ViewBinder<Any, RecyclerView.ViewHolder>>
    ) : ItemMatcher {

        override fun requestViewType(date: Any): Int {
            val dataType = date.javaClass
            return dataTypeToIndex[dataType]
                    ?: throw  RuntimeException("Not found match item, make sure it has been registered: ${dataType.name}")
        }

        override fun requestViewHolderBinder(viewType: Int): ViewBinder<Any, RecyclerView.ViewHolder> {
            return indexToBinder[viewType]
        }
    }
}